modelon/impact/client/client.py

```diff
@@ -77,8 +77,15 @@ def __init__(
                 modelon.impact.client.credential_manager.CredentialManager()
             )
 
+        def credential_resolver():
```

Comment:
I am unsure about the readability of this closure and how that mixes in within the need for the complex setattr decorator pattern - is there something that can be done to make simpler to reason about?

Reply:
The only alternative I could think of would be to have a decorator lik `retry_with_login_decorator` but have it get access to the `credential_resolver` through `self` and then add this decorator to every method of all service classes.

This solution would be more explicit. The downside is that the credential_resolver functions would be passed around everywhere, and you need to remember to add the decorator to all methods.

Reply:
could move all the of the _authenticate_against_api functionality into the service class and mirror the same style as modelon.impact.client.jupyterhub.authorize - in addition the call to api_login which lives in the service class is part of the instance method func wrapper being passed around.

Reply:
I belived I have simplified it with less implicit dependencies. Please have a look.